### PR TITLE
Apache Spark plugin

### PR DESCRIPTION
**Pull request by ShubhamDX**
### Required for all PRs:
- [x] CHANGELOG.md updated
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -37,7 +37,6 @@ prepare:
 # Use the windows godeps file to prepare dependencies
 prepare-windows:
 	go get github.com/sparrc/gdm
-	gdm restore
```
Comment (sparrc) [inline Makefile]: why did you delete this?

File: plugins/inputs/spark/spark.go [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+		fmt.Printf("Missing the required key in output response\n%v\n", out)
+		return
+	}
+
+}
+
+func (j *Spark) SampleConfig() string {
+	return `
+  ## Spark server exposing jolokia read service
+  SparkServer = ["127.0.0.1:8778"] #optional
```
Comment (sparrc) [inline plugins/inputs/spark/spark.go]: this should be `spark_server`

File: plugins/inputs/spark/spark.go [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+  ## Server running Yarn Resource Manager
+  YarnServer = "127.0.0.1:8088" #optional
```
Comment (sparrc) [inline plugins/inputs/spark/spark.go]: this should be `yarn_server`

Comment (sparrc) [inline plugins/inputs/spark/spark.go]: why is SparkServer an array and YarnServer is a string?

File: plugins/inputs/spark/spark.go [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+	tokens := parseJmxMetricRequest(mbean)
+	addTokensToTags(tokens, tags)
+	tags["spark_host"] = c.host
+
+	addValuesAsFields(values, fields, tags["mname"])
+	c.acc.AddFields(tokens["class"]+tokens["type"], fields, tags)
+
+}
+
+func (c sparkMetric) addTagsFields(out map[string]interface{}) {
```
Comment (sparrc) [inline plugins/inputs/spark/spark.go]: why is this a value receiver?

File: plugins/inputs/spark/spark.go [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+			if v != nil {
+				fields[k] = v
+			}
+		}
+	}
+	c.acc.AddFields(class, fields, tags)
+
+}
+
+func (j javaMetric) addTagsFields(out map[string]interface{}) {
```
Comment (sparrc) [inline plugins/inputs/spark/spark.go]: value receiver?

**Comment by sparrc** [comment]
@ShubhamDX why can't spark users just use the `jolokia` plugin? What functionality does this add beyond that?

File: plugins/inputs/spark/spark.go [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+	fmt.Println(out["request"])
```
Comment (sparrc) [inline plugins/inputs/spark/spark.go]: delete these debugging println lines

File: plugins/inputs/spark/spark.go (hunk 5 above)
Comment (sparrc) [inline plugins/inputs/spark/spark.go]: value receiver?

File: plugins/inputs/spark/spark.go (hunk 4 above)
Comment (sparrc) [inline plugins/inputs/spark/spark.go]: value receiver?

File: plugins/inputs/spark/README.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
+# Telegraf plugin: Spark
+
+#### Plugin arguments:
+- **SparkServer** []string: List of spark nodes with the format ["host:port"] (optional)
+- **YarnServer** string: Address of Yarn resource manager (optional)
+
+#### Description
+
+The Spark plugin collects metrics in 2 ways,both being optional: <br />
```
Comment (sparrc) [inline plugins/inputs/spark/README.md]: `<br />` is not necessary, use an actual line break.

Comment (sparrc) [inline plugins/inputs/spark/README.md]: space after `,`

File: plugins/inputs/spark/spark.go (hunk 3 above)
Comment (ShubhamDX) [inline plugins/inputs/spark/spark.go]: Yarn is a central resource manager for the spark cluster. While spark cluster may have many nodes.

File: plugins/inputs/spark/spark.go (hunk 4 above)
Comment (ShubhamDX) [inline plugins/inputs/spark/spark.go]: Didn't get what you said.

**Comment by ShubhamDX** [comment]
Using jolokia will give only JVM stats of that very spark node. While it's resource-manager will give cluster wide information as well as detailed information of each and every application. I have included jolokia so that user should not have to bother about jolokia conf and it's mbeans separately.

File: plugins/inputs/spark/spark.go (hunk 4 above)
Comment (sparrc) [inline plugins/inputs/spark/spark.go]: why is this not `func (c *sparkMetric)`?

File: Makefile (hunk 1 above)
Comment (ShubhamDX) [inline Makefile]: My mistake. Restored.

File: plugins/inputs/spark/spark.go (hunk 2 above)
Comment (ShubhamDX) [inline plugins/inputs/spark/spark.go]: done

File: plugins/inputs/spark/spark.go (hunk 3 above)
Comment (ShubhamDX) [inline plugins/inputs/spark/spark.go]: YarnServer renamed to yarn_server

File: plugins/inputs/spark/spark.go (hunk 6 above)
Comment (ShubhamDX) [inline plugins/inputs/spark/spark.go]: deleted

File: plugins/inputs/spark/README.md (hunk 7 above)
Comment (ShubhamDX) [inline plugins/inputs/spark/README.md]: Done

File: plugins/inputs/spark/spark.go (hunk 4 above)
Comment (ShubhamDX) [inline plugins/inputs/spark/spark.go]: It doesn't matter. Still for best practices I have converted it to pointer receiver

File: plugins/inputs/spark/spark.go (hunk 5 above)
Comment (ShubhamDX) [inline plugins/inputs/spark/spark.go]: Done the changes.

**Comment by sparrc** [comment]
@ShubhamDX can you normalize the metric names to fit better with the general style of telegraf (snakecase)? ie, `spark_HeapMemoryUsage` should be changed to `spark_heap_memory_usage`

you might find the internal.SnakeCase function useful (https://github.com/influxdata/telegraf/blob/master/internal/internal.go#L150)

**Comment by ShubhamDX** [comment]
Done

**Comment by ShubhamDX** [comment]
@sparrc Any update?

**Comment by ShubhamDX** [comment]
@sparrc When can i expect this to be merged into mainline?

**Comment by sparrc** [comment]
I'd give it a few weeks, as you can see there is a bit of a PR backlog at the moment

**Comment by burdandrei** [comment]
wow, this could be really useful,  any news when this will go in the mainline?

**Comment by sparrc** [comment]
@burdandrei it probably won't, we are currently working on a system for having external plugins and it will go there first.

**Comment by ShubhamDX** [comment]
@burdandrei It has been pending for a long. @sparrc can give the required update of going this into mainline

**Comment by burdandrei** [comment]
@sparrc wow, sounds interesting, interesting how it'll be defferent from simple exec plugin.

**Comment by phemmer** [comment]
Maybe we should qualify the name a little bit more. There are several projects called "spark". There's [Cisco Spark](https://www.ciscospark.com/), [particle (formerly known as spark)](https://github.com/spark/), [apache spark](spark.apache.org), [spark IM](https://igniterealtime.org/projects/spark/), and I'm sure others. We use the Cisco Spark, and so when I saw this I was like "sweet, we can monitor Spark", but then after looking into it, it turned into "oh, wrong spark".

**Comment by glinton** [comment]
Closing due to lack of activity, will continue in 3825. @ShubhamDX if you are still interested in addressing the feedback, I can reopen this.
